pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -607,36 +614,48 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
 // determineSchedulingHintForInFlightPod looks at the unschedulable plugins of the given Pod
 // and determines the scheduling hint for this Pod while checking the events that happened during in-flight.
 func (p *PriorityQueue) determineSchedulingHintForInFlightPod(logger klog.Logger, pInfo *framework.QueuedPodInfo, podSchedulingCycle int64) framework.QueueingHint {
-	if len(pInfo.UnschedulablePlugins) == 0 {
-		// When there is no unschedulable plugin, we cannot have a guess which event makes this Pod schedulable.
-		// Here, we use the latest requestCycle so that this Pod won't be stuck in the unschedulable pod pool for a long time.
-		if p.receivedEvents.Len() != 0 {
+	logger.V(5).Info("Checking events for in-flight pod", "pod", klog.KObj(pInfo.Pod), "unschedulablePlugins", pInfo.UnschedulablePlugins, "inFlightEventsSize", p.inFlightEvents.Len(), "inFlightPodsSize", len(p.inFlightPods))
+
+	// AddUnschedulableIfNotPresent is called with the Pod at the end of scheduling or binding.
+	// So, given pInfo should have been Pop()ed before,
+	// we can assume pInfo must be recorded in inFlightPods and thus inFlightEvents.
+	inFlightPod, ok := p.inFlightPods[pInfo.Pod.UID]
+	if !ok {
+		// This can happen while updating a pod. In that case pInfo.UnschedulablePlugins should
+		// be empty. If it is not, we may have a problem.
+		if len(pInfo.UnschedulablePlugins) != 0 {
+			logger.Error(nil, "In flight Pod isn't found in the scheduling queue. If you see this error log, it's likely a bug in the scheduler.", "pod", klog.KObj(pInfo.Pod))
```

Comment:
Code coverage and dlv show that this line is hit when running tests, for example:
```
  2669:				for _, op := range test.operations {
  2670:					for _, pInfo := range pInfos {
=>2671:						op(logger, queue, pInfo)
  2672:					}
  2673:				}
```

This is another instance of "`pInfo` is not the result of `Pop`". Doesn't look easy to fix, though.

Reply:
This code hits it reliably: https://github.com/kubernetes/kubernetes/blob/7e9fbc449ddccab5be18d7d5fa0a4158ec8227f2/pkg/scheduler/internal/queue/scheduling_queue_test.go#L2655

That's `addUnschedulablePodBackToBackoffQ` called for a fabricated `QueuedPodInfo`, without making sure that "inFlightPods" is set up. 

Current master also hits the error log message. This seems to be the only test which does, so it might be easier to fix than I thought.

Reply:
ah I see. Yeah, we should make sure to add a Pop before calling unschedulable. I remember Kensei added a lot of those in several tests. These probably slipped.